Regression in key-project after checker framework update: 3.42-0-eisop5 to 3.49.1-eisop1

The latest update for the checkerFramework (bumped by @dependabot) resulted in regression at the KeY project. [see KeY PR 3589](https://github.com/KeYProject/key/pull/3589)

## Update was: 

| Package | From | To |
| --- | --- | --- |
| [ch.qos.logback:logback-classic](https://github.com/qos-ch/logback) | `1.5.17` | `1.5.18` |
| [io.github.eisop:checker-qual](https://github.com/eisop/checker-framework) | `3.42.0-eisop5` | `3.49.1-eisop1` |
| [io.github.eisop:checker-util](https://github.com/eisop/checker-framework) | `3.42.0-eisop5` | `3.49.1-eisop1` |
| [io.github.eisop:checker](https://github.com/eisop/checker-framework) | `3.42.0-eisop5` | `3.49.1-eisop1` |
| org.checkerframework | `0.6.52` | `0.6.53` |


## Regression

All tests, including the checker framework, passed on the pipeline for the branch, but now the checker framework fails on `main`. 

Errors on main are (see also [Github action log](https://github.com/KeYProject/key/actions/runs/14208137940/job/39810128249)):

```
/home/runner/work/key/key/key.util/src/main/java/org/key_project/util/collection/Immutables.java:152: error: [type.arguments.not.inferred] Could not infer type arguments for DefaultImmutableSet.fromImmutableList
        return DefaultImmutableSet.fromImmutableList(createListFrom(iterable));
                                                    ^
  unsatisfiable constraint: T extends @Nullable Object <: @NonNull Object T extends @Nullable Object <: @NonNull Object
/home/runner/work/key/key/key.util/src/main/java/org/key_project/util/collection/ImmutableArray.java:162: error: [type.arguments.not.inferred] Could not infer type arguments for ArrayIterator constructor
        return new ArrayIterator<>(this);
               ^
  unsatisfiable constraint: S extends @Nullable Object <: @NonNull Object
```

These can be resolved by adding [more annotations](https://github.com/KeYProject/key/commit/5a29d33ae0663e1cd6bd3aa10d55a764b6bda02f). 


Adding these annotations results in the following errors in the key.util test cases:

```
Note: The Checker Framework is tested with JDK 8, 11, 17, and 21. You are using version 23.
Note: Checker Framework 3.49.1-eisop1
/home/weigl/work/key/key.util/src/test/java/org/key_project/util/collection/TestImmutables.java:152: error: [assignment.type.incompatible] incompatible types in assignment.
        ImmutableSet<Object> s1 = DefaultImmutableSet.nil().add("1");
                                                               ^
  found   : @NonNull ImmutableSet<@Nullable Object>
  required: @Nullable ImmutableSet<@NonNull Object>
/home/weigl/work/key/key.util/src/test/java/org/key_project/util/collection/TestImmutables.java:153: error: [assignment.type.incompatible] incompatible types in assignment.
        ImmutableSet<Object> s2 = DefaultImmutableSet.nil().add("2");
                                                               ^
  found   : @NonNull ImmutableSet<@Nullable Object>
  required: @Nullable ImmutableSet<@NonNull Object>
/home/weigl/work/key/key.util/src/test/java/org/key_project/util/collection/TestImmutables.java:163: error: [assignment.type.incompatible] incompatible types in assignment.
        ImmutableSet<Object> s1 = DefaultImmutableSet.nil().add("one").add("two");
                                                                          ^
  found   : @NonNull ImmutableSet<@Nullable Object>
  required: @Nullable ImmutableSet<@NonNull Object>
/home/weigl/work/key/key.util/src/test/java/org/key_project/util/collection/TestImmutables.java:164: error: [assignment.type.incompatible] incompatible types in assignment.
        ImmutableSet<Object> s2 = DefaultImmutableSet.nil().add("two").add("one");
                                                                          ^
  found   : @NonNull ImmutableSet<@Nullable Object>
  required: @Nullable ImmutableSet<@NonNull Object>
```

These might be fixable by playing around. More critical are the NPEs: 

```
/home/weigl/work/key/key.util/src/test/java/org/key_project/util/testcase/java/CollectionUtilTest.java:115: error: [type.argument.inference.crashed] Type argument inference crashed for CollectionUtil.searchAndRemoveWithException
            CollectionUtil.searchAndRemoveWithException(collection, element -> {
                                                       ^
  error: An exception occurred: Cannot invoke "javax.lang.model.type.TypeMirror.getKind()" because "t" is null
/home/weigl/work/key/key.util/src/test/java/org/key_project/util/testcase/java/CollectionUtilTest.java:115: error: [type.argument.inference.crashed] Type argument inference crashed for CollectionUtil.searchAndRemoveWithException
            CollectionUtil.searchAndRemoveWithException(collection, element -> {
                                                       ^
  error: An exception occurred: Cannot invoke "javax.lang.model.type.TypeMirror.getKind()" because "t" is null
/home/weigl/work/key/key.util/src/test/java/org/key_project/util/testcase/java/CollectionUtilTest.java:115: error: [type.argument.inference.crashed] Type argument inference crashed for CollectionUtil.searchAndRemoveWithException
            CollectionUtil.searchAndRemoveWithException(collection, element -> {
                                                       ^
  error: An exception occurred: Cannot invoke "javax.lang.model.type.TypeMirror.getKind()" because "t" is null
```

PR for preparation of the fix: https://github.com/KeYProject/key/pull/3591

/cc @WolframPfeifer @mattulbrich

--- a/framework/src/main/java/org/checkerframework/framework/util/typeinference8/util/CheckedExceptionsUtil.java
+++ b/framework/src/main/java/org/checkerframework/framework/util/typeinference8/util/CheckedExceptionsUtil.java
@@ -73,7 +73,7 @@ public class CheckedExceptionsUtil {
             if (result == null) {
                 result = new ArrayList<>();
             }
-            TypeMirror type = TreeUtils.typeOf(node);
+            TypeMirror type = TreeUtils.typeOf(node.getExpression());
             if (isCheckedException(type, context)) {
                 result.add(type);
             }
@@ -162,7 +162,7 @@ public class CheckedExceptionsUtil {
      */
     private static boolean isCheckedException(TypeMirror type, Java8InferenceContext context) {
         TypeMirror runtimeEx = context.runtimeEx;
-        return context.env.getTypeUtils().isSubtype(type, runtimeEx);
+        return !context.env.getTypeUtils().isSubtype(type, runtimeEx);
     }
 
     /**
@@ -215,7 +215,7 @@ public class CheckedExceptionsUtil {
             if (result == null) {
                 result = new ArrayList<>();
             }
-            AnnotatedTypeMirror type = context.typeFactory.getAnnotatedType(node);
+            AnnotatedTypeMirror type = context.typeFactory.getAnnotatedType(node.getExpression());
             if (isCheckedException(type, context)) {
                 result.add(type);
             }
@@ -314,6 +314,6 @@ public class CheckedExceptionsUtil {
     private static boolean isCheckedException(
             AnnotatedTypeMirror type, Java8InferenceContext context) {
         TypeMirror runtimeEx = context.runtimeEx;
-        return context.env.getTypeUtils().isSubtype(type.getUnderlyingType(), runtimeEx);
+        return !context.env.getTypeUtils().isSubtype(type.getUnderlyingType(), runtimeEx);
     }
 }

--- a/framework/tests/all-systems/java8inference/ExceptionInLambdaTest.java
+++ b/framework/tests/all-systems/java8inference/ExceptionInLambdaTest.java
@@ -1,0 +1,21 @@
+public final class ExceptionInLambdaTest {
+
+    public static void exceptionInLambda() {
+        try {
+            runLambda(new Object(), obj -> {
+                throw new Exception();
+            });
+        } catch (Exception e) {
+        }
+    }
+
+    public static <T, E extends Throwable> void runLambda(
+            T t, ConsumerWithException<T, E> lambda) throws E {
+        lambda.run(t);
+    }
+
+    public static interface ConsumerWithException<T, E extends Throwable> {
+
+        void run(T element) throws E;
+    }
+}

--- a/framework/tests/all-systems/java8inference/ExceptionInLambdaTest.java
+++ b/framework/tests/all-systems/java8inference/ExceptionInLambdaTest.java
@@ -2,15 +2,17 @@ public final class ExceptionInLambdaTest {
 
     public static void exceptionInLambda() {
         try {
-            runLambda(new Object(), obj -> {
-                throw new Exception();
-            });
+            runLambda(
+                    new Object(),
+                    obj -> {
+                        throw new Exception();
+                    });
         } catch (Exception e) {
         }
     }
 
-    public static <T, E extends Throwable> void runLambda(
-            T t, ConsumerWithException<T, E> lambda) throws E {
+    public static <T, E extends Throwable> void runLambda(T t, ConsumerWithException<T, E> lambda)
+            throws E {
         lambda.run(t);
     }
 

